hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/AbstractManagedParentQueue.java

```diff
@@ -173,50 +168,31 @@ public AutoCreatedLeafQueueConfig getLeafQueueTemplate() {
     return queueManagementPolicy;
   }
 
-  protected SortedMap<String, String> getConfigurationsWithPrefix
-      (SortedMap<String, String> sortedConfigs, String prefix) {
-    return sortedConfigs.subMap( prefix, prefix + Character.MAX_VALUE );
-  }
-
-  protected SortedMap<String, String> sortCSConfigurations() {
-    SortedMap<String, String> sortedConfigs = new TreeMap(
-        new Comparator<String>() {
-          public int compare(String s1, String s2) {
-            return s1.compareToIgnoreCase(s2);
-          }
+  protected Map<String, String> getCSConfigurationsWithPrefix(String prefix) {
+    Map<String, String> configsWithPrefix = new HashMap<>();
 
-        });
-
-    for (final Iterator<Map.Entry<String, String>> iterator =
-         csContext.getConfiguration().iterator(); iterator.hasNext(); ) {
-      final Map.Entry<String, String> confKeyValuePair = iterator.next();
-      sortedConfigs.put(confKeyValuePair.getKey(), confKeyValuePair.getValue());
+    for (Map.Entry<String, String> confKeyValuePair :
+        csContext.getConfiguration().getPropsWithPrefix(prefix).entrySet()) {
+      configsWithPrefix.put(prefix + confKeyValuePair.getKey(), confKeyValuePair.getValue());
     }
-    return sortedConfigs;
+
+    return configsWithPrefix;
   }
 
-  protected CapacitySchedulerConfiguration initializeLeafQueueConfigs(String
-      configPrefix) {
+  protected CapacitySchedulerConfiguration initializeLeafQueueConfigs(String configPrefix) {
 
     CapacitySchedulerConfiguration leafQueueConfigs = new
-        CapacitySchedulerConfiguration(new Configuration(false), false);
+        CapacitySchedulerConfiguration(csContext.getConf(), false);
```

Comment:
This isn't entirely what we had in mind. Now every time a new dynamic queue is created every single property of capacity-scheduler.xml and yarn-site.xml is loaded into the newly created leafQueueConfigs, which could hinder the performance. The jira was more about getting rid of the config object duplication in the first place.

Reply:
Hi @brumi1024 , thanks for your review. For my understanging, the `AbstractManagedParentQueue#initializeLeafQueueConfigs` is used to only initialize the leafQueueTemplate when the parent queue is initialized/re-initialized. And for each time, a newly created leafQueue should use this `leafQueueTemplate` to get its configuration via adding new configurations by replacing configurations with `leaf-queue-template`.
But now, [in `ManagedParentQueue#getLeafQueueConfigs`](https://github.com/apache/hadoop/blob/24af52efaba8308e87d1790ccd915c594cf760cc/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/ManagedParentQueue.java#L466), it creates a new configuration everytime. So we need to reuse `getLeafQueueTemplate().getLeafQueueConfigs()` directly instead of creating a new one. I mean we can change this to
```
public CapacitySchedulerConfiguration getLeafQueueConfigs(
      CapacitySchedulerConfiguration templateConfig, String leafQueueName) {
    CapacitySchedulerConfiguration leafQueueConfigTemplate = templateConfig;
    .....
  }
```
What do you think of?

Reply:
Hi @JackWangCS,

Sorry I was not entirely clear. By now I mean the currently merged code creates a new Configuration object in ManagedParentQueue.getLeafQueueConfigs(CapacitySchedulerConfiguration templateConfig, String leafQueueName) (which is called every time an AutoCreatedLeafQueue is instantiated), but your patch solves that issue. However I'm thinking about a bigger change by completely removing the config object duplication (because it is error prone, see the bugfix YarnConfiguration.RESOURCE_TYPES property cloning) by doing something similar which the AQCv2 does (see the AutoCreatedQueueTemplate variable and its uses in ParentQueue and LeafQueue). This way the new template configs are added as if they would be normal user-defined configs, but it's better than creating a new half-filled config for every AutoCreatedLeafQueue or after your patch ManagedParent.

Reply:
I agree with @brumi1024 on this. We need to eliminate initializing an other configuration object, which is still happening in your current revision. My advise is to simply use csContext.getConfiguration to set the properties similar to AutoCreatedQueueTemplate class, as @brumi1024 suggested.

Reply:
Thanks @9uapaw and @brumi1024. I made the initializeLeafQueueConfigs to use the csContext.getConfiguration directly. I will check the AQCv2 related code to see where we can reduce these configuration object duplication further.